File: backend/src/schemas/worker.py
"""
Worker Schemas

Request and response models for worker endpoints.
"""

from datetime import datetime
from typing import Optional, Any, Dict, List, Literal
from uuid import UUID

from pydantic import BaseModel, Field


class WorkerRegister(BaseModel):
    """Worker registration request."""

    machine_id: str = Field(..., min_length=1, max_length=100)
    machine_name: str = Field(..., min_length=1, max_length=100)
    tools: List[str] = Field(default_factory=list)
    system_info: Optional[Dict[str, Any]] = None


class WorkerHeartbeat(BaseModel):
    """Worker heartbeat request."""

    status: str = Field(..., pattern=r"^(online|offline|busy|idle)$")
    cpu_percent: Optional[float] = Field(None, ge=0, le=100)
    memory_percent: Optional[float] = Field(None, ge=0, le=100)
    disk_percent: Optional[float] = Field(None, ge=0, le=100)
    tools: Optional[List[str]] = None
    current_task_id: Optional[UUID] = None


class WorkerResponse(BaseModel):
    """Worker response."""

    worker_id: UUID
    machine_id: str
    machine_name: str
    status: str
    tools: List[str]
    system_info: Optional[Dict[str, Any]] = None
    cpu_percent: Optional[float] = None
    memory_percent: Optional[float] = None
    disk_percent: Optional[float] = None
    last_heartbeat: Optional[datetime] = None
    registered_at: datetime

    model_config = {"from_attributes": True}


class WorkerListResponse(BaseModel):
    """Paginated worker list response."""

    workers: List[WorkerResponse]
    total: int
    limit: int
    offset: int


class WorkerTaskAssignment(BaseModel):
    """Task assignment pushed to worker."""

    task_id: UUID
    description: str
    tool_preference: Optional[str] = None
    priority: int
    workflow_id: Optional[UUID] = None
    metadata: Optional[Dict[str, Any]] = None


class TaskCompleteRequest(BaseModel):
    """Task completion request from worker."""

    task_id: UUID
    result: Dict[str, Any] = Field(default_factory=dict)


class TaskFailedRequest(BaseModel):
    """Task failure request from worker."""

    task_id: UUID
    error: str


class TaskResultReport(BaseModel):
    """Task result report from worker - unified completion/failure reporting."""

    task_id: UUID
    status: Literal["completed", "failed", "cancelled"]
    result: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    execution_time_ms: int = Field(..., ge=0, description="Execution time in milliseconds")
    metrics: Optional[Dict[str, int]] = Field(
        None, description="Execution metrics (e.g., tokens_used, api_calls)"
    )


class TaskResultResponse(BaseModel):
    """Response for task result reporting."""

    status: str = "success"
    task_id: UUID
    task_status: str
    message: str

File: frontend-v2/src/types/worker.ts
/**
 * Worker Types
 *
 * TypeScript type definitions for worker-related data structures.
 */

// Worker status enum
export type WorkerStatus = 'online' | 'offline' | 'busy' | 'idle';

// Available AI tools
export type ToolName = 'claude_code' | 'gemini_cli' | 'ollama' | string;

// Tool status
export type ToolStatus = 'available' | 'busy' | 'unavailable' | 'error';

// System info structure
export interface SystemInfo {
  os?: string;
  cpu?: string;
  memory_total_gb?: number;
  disk_total_gb?: number;
  platform?: string;
  architecture?: string;
  [key: string]: string | number | undefined;
}

// Worker interface matching backend WorkerResponse
export interface Worker {
  worker_id: string;
  machine_id: string;
  machine_name: string;
  status: WorkerStatus;
  tools: ToolName[];
  system_info?: SystemInfo | null;
  cpu_percent?: number | null;
  memory_percent?: number | null;
  disk_percent?: number | null;
  last_heartbeat?: string | null;
  registered_at: string;
  is_active?: boolean;
}

// Worker registration request
export interface WorkerRegisterRequest {
  machine_id: string;
  machine_name: string;
  tools: ToolName[];
  system_info?: SystemInfo | null;
}

// Worker heartbeat request
export interface WorkerHeartbeatRequest {
  status: WorkerStatus;
  cpu_percent?: number | null;
  memory_percent?: number | null;
  disk_percent?: number | null;
  tools?: ToolName[] | null;
  current_task_id?: string | null;
}

// Worker list response
export interface WorkerListResponse {
  workers: Worker[];
  total: number;
  limit: number;
  offset: number;
}

// Worker filter options
export interface WorkerFilters {
  status?: WorkerStatus | 'all';
  tools?: ToolName[];
  search?: string;
}

// Worker sort options
export type WorkerSortField = 'machine_name' | 'status' | 'last_heartbeat' | 'registered_at';
export type SortDirection = 'asc' | 'desc';

export interface WorkerSortOptions {
  field: WorkerSortField;
  direction: SortDirection;
}

// Worker metrics history point
export interface MetricsDataPoint {
  timestamp: string;
  cpu_percent: number;
  memory_percent: number;
  disk_percent: number;
}

// Worker metrics summary
export interface WorkerMetricsSummary {
  current_cpu: number;
  current_memory: number;
  current_disk: number;
  avg_cpu_24h?: number;
  avg_memory_24h?: number;
  task_completion_rate?: number;
  avg_execution_time_ms?: number;
  total_tasks_completed?: number;
  total_tasks_failed?: number;
}

// Worker task history item
export interface WorkerTaskHistoryItem {
  task_id: string;
  description: string;
  status: string;
  priority: number;
  tool_preference?: string | null;
  created_at: string;
  completed_at?: string | null;
  execution_time_ms?: number | null;
}

// Worker form data for create/edit
export interface WorkerFormData {
  machine_name: string;
  description?: string;
  tools: ToolName[];
  is_active: boolean;
}

// Tool with status info
export interface ToolInfo {
  name: ToolName;
  status: ToolStatus;
  version?: string;
  last_used?: string;
}

// View mode for worker list
export type ViewMode = 'grid' | 'table';

// Bulk action types
export type BulkAction = 'activate' | 'deactivate' | 'delete';
